### PR TITLE
476 toggle permission to uploaddelete update move copy and rename

### LINKED ISSUE #476: toggle permission to  `upload`,`delete`, `update`, `move`, `copy` and `rename`.
There is a new toggle to allow  `upload`,`delete`, `update`, `move`, `copy` and `rename`.

Boolean fields controlling this will be added to the allocation object blobber gets from 0cahin. Each of `upload`,`delete`, `update`, `move`, `copy` and `rename` should only be permitted if the corresponding field in the allocation object permits it.

This change is blocked by https://github.com/0chain/0chain/issues/828.

### PR DESCRIPTION
**Pull request by devyetii**
### Changes
- Changes regarding allocation to support NFT by adding `file_options` to guard object operations on this allocation

### Fixes
#476 

### Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/blobber/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

### Associated PRs (Link as appropriate):
- 0chain: 0chain/0chain#828
- gosdk: 0chain/gosdk#686 **Required**
- system_test:
- zboxcli: 0chain/zboxcli#149
- zwalletcli:
- Other: ...

### CODE REVIEW

**Comment by devyetii** [comment]
requires 0chain/gosdk#687 to be merged

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | 686-add-fields-to-allocation-to-support-nft  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | 476-toggle-permission-to-uploaddelete-update-move-copy-and-rename  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 398-new-system-test-for-nft-allocation-features  |
| gosdk  | staging  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | 476-toggle-permission-to-uploaddelete-update-move-copy-and-rename  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 398-new-system-test-for-nft-allocation-features  |
| gosdk  | 686-add-fields-to-allocation-to-support-nft  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | 476-toggle-permission-to-uploaddelete-update-move-copy-and-rename  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 398-new-system-test-for-nft-allocation-features  |
| gosdk  | 686-add-fields-to-allocation-to-support-nft  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | 476-toggle-permission-to-uploaddelete-update-move-copy-and-rename  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | 686-add-fields-to-allocation-to-support-nft  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | 476-toggle-permission-to-uploaddelete-update-move-copy-and-rename  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | 686-add-fields-to-allocation-to-support-nft  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | 476-toggle-permission-to-uploaddelete-update-move-copy-and-rename  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 398-new-system-test-for-nft-allocation-features  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | 476-toggle-permission-to-uploaddelete-update-move-copy-and-rename  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/blobbercore/allocation/entity.go [hunk 1]
```diff
@@ -49,6 +55,21 @@ type Allocation struct {
 	// Ending and cleaning
 	CleanedUp bool `gorm:"column:cleaned_up;not null;default:false"`
 	Finalized bool `gorm:"column:finalized;not null;default:false"`
+
+	// Flag to determine if anyone can extend this allocation
+	ThirdPartyExtendable bool `json:"third_party_extendable" gorm:"column:third_party_extendable;not null;default:false"`
```
Comment (dabasov) [inline code/go/0chain.net/blobbercore/allocation/entity.go]: not sure that we need this field on blobber

File: code/go/0chain.net/blobbercore/handler/object_operation_handler.go [hunk 2]
```diff
@@ -539,6 +539,10 @@ func (fsh *StorageHandler) RenameObject(ctx context.Context, r *http.Request) (i
 		return nil, common.NewError("immutable_allocation", "Cannot rename data in an immutable allocation")
 	}
 
+	if ! allocationObj.CanRename() {
```
Comment (dabasov) [inline code/go/0chain.net/blobbercore/handler/object_operation_handler.go]: please remove space between `!` and `allocation`

**Review by dabasov — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/blobbercore/allocation/entity.go (hunk 1 above)
Comment (devyetii) [reply code/go/0chain.net/blobbercore/allocation/entity.go]: It's actually not used so yes I'll be removing that.

**Review by dabasov — APPROVED** [review]
lgtm

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 398-new-system-test-for-nft-allocation-features  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | 476-toggle-permission-to-uploaddelete-update-move-copy-and-rename  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Pull request merged**